File: copilot-silent-chat-vscode/src/orchestrator/ChatOrchestrator.ts
import * as vscode from 'vscode';
import { CopilotSilentChatService } from '../service/CopilotSilentChatService';
import { SilentChatEvent } from '../model/events';
import { randomUUID } from 'crypto';

/**
 * ChatOrchestrator — equivalent of ChatOrchestrator.kt
 *
 * Dispatches multiple messages in sequence or parallel,
 * coordinating sessions and collecting results.
 */

export interface OrchestratorMessage {
  message: string;
  model?: string;
  mode?: string;
  sessionId?: string;
}

export interface OrchestratorResult {
  sessionId: string;
  response: string;
  success: boolean;
  error?: string;
}

export class ChatOrchestrator implements vscode.Disposable {
  private disposables: vscode.Disposable[] = [];

  constructor(private readonly chatService: CopilotSilentChatService) {}

  /**
   * Send messages sequentially, each waiting for the previous to complete.
   */
  async sendSequential(messages: OrchestratorMessage[]): Promise<OrchestratorResult[]> {
    const results: OrchestratorResult[] = [];

    for (const msg of messages) {
      const result = await this.sendAndCollect(msg);
      results.push(result);

      // Stop on error
      if (!result.success) {
        break;
      }
    }

    return results;
  }

  /**
   * Send messages in parallel, all starting at once.
   */
  async sendParallel(messages: OrchestratorMessage[]): Promise<OrchestratorResult[]> {
    const promises = messages.map((msg) => this.sendAndCollect(msg));
    return Promise.all(promises);
  }

  /**
   * Send a single message and collect the full response.
   */
  private async sendAndCollect(msg: OrchestratorMessage): Promise<OrchestratorResult> {
    const sessionId = msg.sessionId ?? randomUUID();
    let accumulated = '';
    let success = true;
    let error: string | undefined;

    // Subscribe to events for this session
    const eventPromise = new Promise<void>((resolve) => {
      const disposable = this.chatService.onEvent(([sid, event]) => {
        if (sid !== sessionId) return;

        switch (event.type) {
          case 'Reply':
            accumulated = event.accumulated;
            break;
          case 'Complete':
            disposable.dispose();
            resolve();
            break;
          case 'Error':
            success = false;
            error = event.message;
            disposable.dispose();
            resolve();
            break;
          case 'Cancel':
            success = false;
            error = 'Cancelled';
            disposable.dispose();
            resolve();
            break;
        }
      });
    });

    // Send the message
    await this.chatService.sendMessage({
      message: msg.message,
      sessionId,
      model: msg.model,
      mode: msg.mode,
    });

    // Wait for completion
    await eventPromise;

    return { sessionId, response: accumulated, success, error };
  }

  dispose(): void {
    this.disposables.forEach((d) => d.dispose());
  }
}

File: citi-ai-intellij-plugin/src/main/kotlin/com/github/copilotsilent/orchestrator/ChatOrchestrator.kt
package com.github.copilotsilent.orchestrator

import com.github.copilot.chat.conversation.agent.rpc.command.ChatMode
import com.github.copilot.chat.conversation.agent.rpc.command.CopilotModel
import com.github.copilotsilent.model.SilentChatEvent
import com.github.copilotsilent.model.SilentChatListener
import com.github.copilotsilent.service.CopilotSilentChatService
import com.intellij.openapi.Disposable
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.util.concurrent.atomic.AtomicReference

/**
 * Orchestrates sequential and parallel message dispatching.
 *
 * - sendParallel(): each request gets its own session, all run concurrently
 * - sendSequential(): requests queue on a single session, each waits for the previous to complete
 */
@Service(Service.Level.PROJECT)
class ChatOrchestrator(
    private val project: Project,
) : Disposable {
    private val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    override fun dispose() {
        coroutineScope.cancel()
    }
    private val log = Logger.getInstance(ChatOrchestrator::class.java)

    private val service: CopilotSilentChatService
        get() = project.service<CopilotSilentChatService>()

    /**
     * Send multiple messages in parallel — each gets its own new session.
     */
    fun sendParallel(requests: List<ChatRequest>) {
        for (request in requests) {
            service.sendMessage(
                message = request.message,
                model = request.model,
                mode = request.mode,
                newSession = true,
                silent = request.silent,
            )
        }
    }

    /**
     * Send multiple messages sequentially on the same session.
     * Each message waits for the previous to complete before sending.
     *
     * Subscribes once to the MessageBus for the entire chain. Uses an
     * AtomicReference to track the current session ID so the subscriber
     * only reacts to events for the active step's session.
     */
    fun sendSequential(
        requests: List<ChatRequest>,
        sessionId: String? = null,
    ) {
        coroutineScope.launch {
            val currentSid = AtomicReference(sessionId)
            val currentCompletion = AtomicReference<CompletableDeferred<String?>>()

            val conn = project.messageBus.connect()
            conn.subscribe(SilentChatListener.TOPIC, object : SilentChatListener {
                override fun onEvent(eventSessionId: String, event: SilentChatEvent) {
                    val expectedSid = currentSid.get()

                    when (event) {
                        is SilentChatEvent.SessionReady -> {
                            // Accept SessionReady only when we're waiting for a new session
                            // (expectedSid is null) — this locks us to that session
                            if (expectedSid == null) {
                                currentSid.set(event.sessionId)
                            }
                        }
                        is SilentChatEvent.Complete -> {
                            if (eventSessionId == currentSid.get()) {
                                currentCompletion.get()?.complete(currentSid.get())
                            }
                        }
                        is SilentChatEvent.Error -> {
                            if (eventSessionId == currentSid.get()) {
                                currentCompletion.get()?.complete(null)
                            }
                        }
                        else -> {}
                    }
                }
            })

            try {
                for ((index, request) in requests.withIndex()) {
                    val completion = CompletableDeferred<String?>()
                    currentCompletion.set(completion)

                    service.sendMessage(
                        message = request.message,
                        sessionId = currentSid.get(),
                        model = request.model,
                        mode = request.mode,
                        newSession = (currentSid.get() == null && index == 0),
                        silent = request.silent,
                    )

                    val result = completion.await()
                    if (result == null) {
                        log.warn("Sequential chain stopped at request $index due to error")
                        break
                    }
                }
            } finally {
                conn.disconnect()
            }
        }
    }
}

data class ChatRequest(
    val message: String,
    val model: CopilotModel? = null,
    val mode: ChatMode? = null,
    val silent: Boolean = true,
)
